gdbplus/gdb-12.1/gdb/gdb_dep.c

```diff
@@ -2797,3 +2797,114 @@ ca_eval_address(const char* expr)
 {
 	return parse_and_eval_address(expr);
 }
+
+// return true if an unamed union has a data member of given name
+static bool
+union_has_field_name(struct type *type, const char *field_name)
+{
+	type = check_typedef (type);
+	if (type->code() != TYPE_CODE_UNION)
+		return false;
+	for (int i = 0; i < type->num_fields(); i++) {
+		const char *name = type->field(i).name();
+		if (name && *name != '\0') {
+			if (strcmp (field_name, name) == 0)
+				return true;
+		} else if (union_has_field_name(type->field(i).type(), field_name)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static int
+type_field_name2no(struct type *type, const char *field_name)
+{
+	if (type == NULL)
+		return -1;
+
+	type = check_typedef (type);
+
+	for (int n = 0; n < type->num_fields(); n++) {
+		const char* name = type->field(n).name();
+		if (name && *name != '\0') {
+			if (strcmp (field_name, name) == 0)
+				return n;
+		} else if (union_has_field_name(type->field(n).type(), field_name)) {
+			return n;
+		}
+	}
+	return -1;
+}
+
+bool
+ca_get_field_value(struct value *val, const char *fieldname,
+				size_t *data, bool optional)
+{
+	int fieldno;
+	struct value *field_val;
+
+	*data = ULONG_MAX;
+	fieldno = type_field_name2no(value_type(val), fieldname);
+	if (fieldno < 0) {
+		if (optional) {
```

Comment:
With the `std::optional` we might be able to get rid of the the extra parameter `optional`

Reply:
yes. But I will defer this because the function is exposed to files that are part of gdb 9 and 7 builds. They are usually built not with the latest g++, which means the default c++ standard is less than c++17 that is required by `std::optional`.

Reply:
I see. Then it is better to stick to the old way.